### DISCUSSION TITLE
Dummy kernel for simulation

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#19247] [Q&A General] [answered]
Hello!

A mesh for my simulation has several phases (A, B, C, and D).

I want to run a simulation (physics) for only one phase (C).

However, as far as I know, MOOSE requires every phase have a kernel.

I tried to NullKernel for A,B, and D, but it did not converge.

Are there any ways to run a simulation without affecting simulation convergence?

Thanks!

Best regards,
Hokon Kim

**Comment by GiudGiud** [comment, marked as answer]
hello

how ironic I was documenting this kernel last week but did not make the PR yet. I ll make it now. 

Best option is to move the variables to aux variables since you are not solving for them. That will decrease the size of the matrix. 

The NullKernel has a jacobian fill parameter that should have done the trick. Can you try increasing it?
Otherwise I d try with a time derivative kernel, since these variables are effectively solving that equation (dA/dt = 0)

Guillaume

**Reply by GiudGiud** [reply]
Here s the future version of the docs for this
https://mooseframework.inl.gov/docs/PRs/19248/site/source/kernels/NullKernel.html

let s improve it as we figure this one out

**Reply by hokonkim** [reply]
As far as I know,

Variables work with Kernels and AuxVariables work with AuxKernels.

If I move some variables to aux variables, then I might use AuxKernels.

Does Kernels (NullKernel) work with aux variables?

I don't run an application with the input below, yet.

    [Variables]
      [./phi_YSZ]
        block = 'PT_YSZ_TET4 PT_TPB_TET4'
        initial_condition = 0.00000 # (V)
        scaling = 1e4
      [../]
    []

    [Kernels]
      [./vacancyIonicDrift]
        type  = DiffMatKernel
        block = 'PT_YSZ_TET4 PT_TPB_TET4'
        variable  = phi_YSZ
        diff_coef = 'sigma_YSZ'
      [../]

      [./gasDiffusion]
        type  = NullKernel
        block = 'PT_PORE_TET4 PT_TPB_TET4'
        variable  = p_O2
      [../]

      [./holeDrift]
        type  = NullKernel
        block = 'PT_LSM_TET4 PT_TPB_TET4' # when you use previously generated mesh, you need to use LSM block.
        variable  = phi_LSM
      [../]
    []

    [AuxVariables]
      [./p_O2]
        block = 'PT_PORE_TET4 PT_TPB_TET4'
        initial_condition = 0.21 # (atm)
        # scaling = 1e4
      [../]

      [./phi_LSM]
        block = 'PT_LSM_TET4 PT_TPB_TET4' # when you use previously generated mesh, you need to use LSM block.
        initial_condition = 1.03665 # (V) Just for condition: Each subdomain must contain at least one Kernel.
        # scaling = 1e7 # 1e2 for cylinder
      [../]
    []

**Reply by GiudGiud** [reply]
If you move the variable to be an aux-Variable then you dont need a NullKernel. The NullKernel is to avoid a singularity in the non-linear system. If you dont want to do anything to an auxvariable then dont have an auxkernel for it

**Reply by hokonkim** [reply]
My question is "even I move some variables to Aux-variable", still some phases (A, B, and D) does not have Kernels.

This will give me an error like "each subdomain has one Kernel.."

**Reply by GiudGiud** [reply]
oh those are less of a concern. 
Just add "kernel_coverage_check = false" to your [Problem] block.

**Reply by hokonkim** [reply]
I modified part of input file like below.

As far as I know, there are not any issues.

I didn't run it with this input file yet. But, let's see if it works.

    [Variables]
      [./phi_YSZ]
        block = 'PT_YSZ_TET4 PT_TPB_TET4'
        initial_condition = 0.00000 # (V)
        scaling = 1e4
      [../]
    []

    [Kernels]
      [./vacancyIonicDrift]
        type  = DiffMatKernel
        block = 'PT_YSZ_TET4 PT_TPB_TET4'
        variable  = phi_YSZ
        diff_coef = 'sigma_YSZ'
      [../]
    []

    [Problem] # To make it possible to run a simulation without assigning kernel or material property to each phase.
      kernel_coverage_check = false
      material_coverage_check = false
    []

    [BCs]
      [./potentialYSZ_bottom]
        type = DirichletBC
        variable = phi_YSZ
        boundary = 'SF_YSZ_WITH_XMAX'
        value = 0.00000 # (V)
      [../]

      [./potentialYSZ_top]
        type = FunctionDirichletBC
        variable = phi_YSZ
        boundary = 'SF_YSZ_WITH_XMIN'
        function = 'funcPotentialYSZ' # (V)
      [../]
    []
